test/e2e/framework.go

```diff
@@ -367,14 +367,30 @@ func (data *TestData) waitForAntreaDaemonSetPods(timeout time.Duration) error {
 		// first time we get the DaemonSet's Status, we would return immediately instead of
 		// waiting.
 		desiredNumber := int32(clusterInfo.numNodes)
-		if daemonSet.Status.NumberAvailable == desiredNumber &&
-			daemonSet.Status.UpdatedNumberScheduled == desiredNumber {
-			// Success
-			return true, nil
+		if daemonSet.Status.NumberAvailable != desiredNumber ||
+			daemonSet.Status.UpdatedNumberScheduled != desiredNumber {
+			return false, nil
 		}
 
-		// Keep trying
-		return false, nil
+		// Make sure that all antrea-agent Pods are not terminating. This is required because NumberAvailable of
+		// DaemonSet counts Pods even if they are terminating. Deleting antrea-agent Pods directly does not cause the
+		// number decrease if the process doesn't quit immediately, e.g. when the signal is caught by bincover program
+		// and triggers coverage calculation.
+		pods, err := data.clientset.CoreV1().Pods(antreaNamespace).List(context.TODO(), metav1.ListOptions{
+			LabelSelector: "app=antrea,component=antrea-agent",
+		})
+		if err != nil {
+			return false, fmt.Errorf("failed to list antrea-agent Pods: %v", err)
+		}
+		if len(pods.Items) != clusterInfo.numNodes {
+			return false, nil
+		}
+		for _, pod := range pods.Items {
+			if pod.DeletionTimestamp != nil {
```

Comment:
Do you use the check of "DeletionTimestamp != nil" to ensure Pod is not Terminating? Could you add a comments here?
Just confused, why not use "pod.Status.Phase == PodRunning" as a check?

Reply:
Isn't `DeletionTimestamp != nil` self-explaining that the pod is not terminating?
`pod.Status` is populated by system (kubelet for Pod's case) and represents its runtime state. Deleting a resource is normally an asynchronous operation and `Status` won't change by user's API call. You can try deleting a Pod via kubectl and no fields in Pod.Status will change immediately, only DeletionTimestamp is set.

Reply:
So you mean for a "deleting" Pod, its Status might be still Running before completion, but its `DeletionTimestamp` must be be set?

Reply:
Yes, a Pod's phase would still be running before kubelet manages to kill it successfully. `Delete` API call sets `DeletionTimestamp`, which triggers graceful termination. Before checking DeletionTimestamp, the function checks DaemonSet status first, which ensures `NumberAvailable` and `UpdatedNumberScheduled` is desired so the Pods must be running, up-to-date, and available. The case the patch fixes is to the Pods are up-to-date and available, but they were marked for termination.

Reply:
Got it, thanks for explanation.